### DISCUSSION TITLE
Check for element in XML feed

### DISCUSSION THREAD

**Discussion by mcampanella222** [nextgenhealthcare/connect#5279] [Q&A] [answered]
Hello

I am working with a stream of XML data form a channel.

I need to write out certain elements into a file.  Everthing works great until one of the transactions is sent without an element named "merchant_defined_field"  This field is missing on some of the transactions so the code to write out the values fails.

How do I test for  the existence of this field so I can skip the line blow if it is missing data?  The XML stream has multiple transaction in  it from a response message.

**XML feed:**
<?xml version="1.0" encoding="UTF-8"?>
<nm_response>
	<transaction>
		<transaction_id>7168629594</transaction_id>
		<partial_payment_
                <entry_mode>Keyed</entry_mode>
                <merchant_defined_field id="1">MRN99999</merchant_defined_field>
                <merchant_defined_field id="2">APPTNUM202022</merchant_defined_field>
	</transaction>
	<transaction>
		<transaction_id>7168629594</transaction_id>
		<partial_payment_
                <entry_mode>Keyed</entry_mode>
	</transaction>
</nm_response>		

**Code to write out file:**
//loop through response

var i;

for(i=0;i<msg['transaction'].length();i++)
{
file+=msg['transaction'][i]['merchant_defined_field'][0].toString()+",";
file+=msg['transaction'][i]['merchant_defined_field'][1].toString()+",";
}
channelMap.put("FILE",file);

**Comment by mjtrac** [comment, marked as answer]
The approach used by Mirth is a try/catch.  If you were to use drag and drop to create a non-iterating transformer and look at the resulting code, it is as follow:
```
try {
  mapping = msg['transaction'].toString(); // your whole for loop could go here.
} catch (e) {
  mapping = ''; //channelMap.put("FILE","No file; no transaction found.");
}
// deal with mapping
```
A try/catch construct is more or less what the name implies.  Try something, and if it generates an exception, then instead of allowing the exception to bubble up, catch it when it occurs by executing code in the catch block.  In this case, if something is absent, just set it to an empty string.  In your catch, you might want to put a flag value into the FILE field of the channelMap if you find your msg contains no transaction.  Or, better, you might want to create a channelMap entry such as NOFILE which you would test for prior to relying on the value of FILE.

**Reply by mcampanella222** [reply]
Thanks so much for the reply.  I was able to trap the transactions missing those elements and skip over them with a comma!

var i;
var nomerchid = '';

for(i=0;i<msg['transaction'].length();i++)

{

try {
  mapping = msg['transaction'][i]['merchant_defined_field'][0].toString();

} catch (e) {


nomerchid = "true"
file+=msg['transaction'][i]['transaction_id'].toString()+",";
file+=",";

**Reply by pacmano1** [reply]
I susepct you can just do this:

```JS
var file = ''
for each (transaction in msg['transaction']) {
	file+='Transaction:' +  transaction['nm_response']['transaction_id'].toString() + '\n'
	for each (merchant_defined_field in transaction['nm_response']['partial_payment']['merchant_defined_field']) {
   		file += 'Merchant Defined Field:' +  merchant_defined_field.toString() + '\n'	
	}
	file=file.slice(0,-1) // remove extra line feed
	file+='\n'
}
channelMap.put('FILE',file); 
```
Passing this:
```xml
<result>
    <transaction>
        <nm_response>
            <transaction_id>7168629594</transaction_id>
            <partial_payment>
                <entry_mode>Keyed</entry_mode>
                <merchant_defined_field id="1">MRN99999</merchant_defined_field>
                <merchant_defined_field id="2">APPTNUM202022</merchant_defined_field>
                <transaction_id>7168629594</transaction_id>
            </partial_payment>
            <entry_mode>Keyed</entry_mode>
        </nm_response>
    </transaction>
    <transaction>
        <nm_response>
            <transaction_id>7168629594</transaction_id>
            <partial_payment>
                <entry_mode>Keyed</entry_mode>
                <transaction_id>7168629594</transaction_id>
            </partial_payment>
            <entry_mode>Keyed</entry_mode>
        </nm_response>
    </transaction>
    <transaction>
        <nm_response>
            <transaction_id>7168629594</transaction_id>
            <partial_payment>
                <entry_mode>Keyed</entry_mode>
                <transaction_id>7168629594</transaction_id>
            </partial_payment>
            <entry_mode>Keyed</entry_mode>
        </nm_response>
    </transaction>
    <transaction>
        <nm_response>
            <transaction_id>7168629594</transaction_id>
            <partial_payment>
                <entry_mode>Keyed</entry_mode>
                <merchant_defined_field id="1">MRN99992</merchant_defined_field>
                <merchant_defined_field id="2">APPTNUM202021</merchant_defined_field>
                <transaction_id>7168629594</transaction_id>
            </partial_payment>
            <entry_mode>Keyed</entry_mode>
        </nm_response>
    </transaction>
</result>
```
results in:
```
Transaction:7168629594
Merchant Defined Field:MRN99999
Merchant Defined Field:APPTNUM202022
Transaction:7168629595
Transaction:7168629596
Transaction:7168629597
Merchant Defined Field:MRN99992
Merchant Defined Field:APPTNUM202021
```

**Reply by pacmano1** [reply]
@mcampanella222, In the future please use code blocks and highlighting also.
